sql/parser/type_check.go

```diff
@@ -228,6 +231,63 @@ func typeCheckComparisonOp(op ComparisonOp, dummyLeft, dummyRight Datum) (Datum,
 		dummyLeft.Type(), op, dummyRight.Type())
 }
 
+func typeCheckConditionalExpr(expr *ConditionalExpr) (Datum, error) {
+	switch expr.Operator {
+	case If:
+		if len(expr.Exprs) != 3 {
```

Comment:
Added a commit which separates `ConditionalExpr` into `IfExpr`, `NullIfExpr` and `CoalesceExpr`. This made certain bits of the code cleaner, but also resulted in more code. I did not see an easy way to introduce a `ConditionalExpr` interface. Such an interface would have to have a function to return a list of the internal expressions, but then we'd need different code for type checking `If` vs `NullIf` vs `Coalesce`.